_posts/2020-06-01-spring-boot-hazelcast.md

```diff
@@ -0,0 +1,232 @@
+how long a data lives in the cache if it is not required. It is called **time-to-live** (TTL).
+
+In a big enterprise system, there can be a cluster of caches. We have to replicate and synchronize the data in this cluster
+between the caches. It is called **write-through concept**.
+
+## Hazelcast as Distributed Cache
+
+Let's say we have a Spring application, and we want to build in a cache in the application. But we also want to 
+be able to scale this application. It means, when we start for example three instances of the application, they have to
+share the cache to keep the data consistently. This problem can be solved by using a **distributed cache**.
+
+Hazelcast is a distributed in-memory object store and provides many features including TTL, work-through, and scalability.
+We can build a Hazelcast cluster by starting several `Hazelcast` nodes in a net. This node is called a member. 
+Before we go into the topic, how to use Hazelcast with Spring, let's have a look at the topologies of Hazelcast 
+cache cluster.
+
+There are two types of topologies
+ * topology with embedded cache,
+ * client-server topology.
+ 
+## Embedded Cache
+This type of topology means, that every instance of application has an integrated member. 
+
+![Embedded Cache Topology](/assets/img/posts/hazelcast/embedded-cache.png)
+
+In this case, the application and the cache data are running on one node. **When a new cache entry is written in
+the cache, `Hazelcast` takes care about the distribution to other members.** When data should be read from the cache, 
+then this data can be found on the same node, where the application is running.
+
+Let's have a look at how to build a cluster with an embedded `Hazelcast` cache topology and Spring application.
+`Hazelcast` supports many distributed data structures for caching. We will use a `Map` because it provides well known `get`
+and `put` operations. 
+
+First, we have to add the `Hazelcast` dependency. `Hazelcast` is just a Java library, so that can be done very easily.
+
+````groovy
+compile group: 'com.hazelcast', name: 'hazelcast', version: '4.0.1'
+````
+
+Now let's create a cache client for the application.
+
+```java
+@Component
+public class CacheClient {
+
+    public static final String CARS = "cars";
+    private final HazelcastInstance hazelcastInstance
+                            = Hazelcast.newHazelcastInstance();
+
+    public Car put(String number, Car car){
+        IMap<String, Car> map = hazelcastInstance.getMap(CARS);
+        return map.putIfAbsent(number, car);
+    }
+
+    public Car get(String key){
+        IMap<String, Car> map = hazelcastInstance.getMap(CARS);
+        return map.get(key);
+    }
+   
+   // other methods omitted
+
+}
+```
+That's it. Now the application has a distributed cache. The most important part of this code is the creation of
+a cluster member. It happens by calling the method `Hazelcast.newHazelcastInstance()`. The method `getMap()`
```

Comment:
How does Hazelcast know which other distributed Hazelcast nodes to exchange data with?

Reply:
I explained it shortly.